File: src/components/ResIns components/ff.js
import React, { useEffect, useState } from 'react';
import '../../styles/ff.css';

const FamousFoods = () => {
  const [insights, setInsights] = useState(null);
  const [error, setError] = useState(null);

  useEffect(() => {
    const fetchFamousFoods = async () => {
      try {
        const response = await fetch('http://localhost:5000/api/famous-foods');
        if (!response.ok) {
          throw new Error('Failed to fetch famous food insights.');
        }
        const data = await response.json();
        setInsights(data.detailed_famous_foods);
      } catch (error) {
        setError(error.message);
      }
    };

    fetchFamousFoods();
  }, []);

  if (error) {
    return <div>Error: {error}</div>;
  }

  if (!insights) {
    return <div>Loading famous food insights...</div>;
  }

  return (
    <div>
      <h2>Top 50 Famous Foods Insights</h2>
      <table>
        <thead>
          <tr>
            <th>Famous Food</th>
            <th>City</th>
            <th>Restaurant Count</th>
            <th>Average Rating</th>
            <th>Average Cost for Two</th>
          </tr>
        </thead>
        <tbody>
          {insights.map((item, index) => (
            <tr key={index}>
              <td>{item.famous_food || 'Not Available'}</td>
              <td>{item.city || 'Not Available'}</td>
              <td>{item.restaurant_count ?? 'N/A'}</td>
              <td>{item.avg_rating ? item.avg_rating.toFixed(2) : 'No Rating'}</td>
              <td>{item.avg_cost_for_two ? item.avg_cost_for_two.toFixed(2) : 'N/A'}</td>
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
};

export default FamousFoods;

File: flask/ff.py
import pandas as pd


def get_famous_food_insights():
    # Load the dataset
    df = pd.read_csv('static/data/rs.csv')

    # Remove rows with missing famous food data
    df = df.dropna(subset=['famous_food'])

    # Insights: Top 10 Famous Foods by Frequency
    top_famous_foods = (
        df['famous_food']
        .value_counts()
        .head(50)
        .reset_index()
    )
    top_famous_foods.columns = ['famous_food', 'restaurant_count']

    # Include city with the highest count for each famous food
    top_famous_foods_cities = df[df['famous_food'].isin(top_famous_foods['famous_food'])]
    popular_food_cities = (
        top_famous_foods_cities.groupby('famous_food')['city']
        .apply(lambda x: x.mode().iloc[0])
        .reset_index()
    )

    # Merge to add city information to the top 10 foods
    top_famous_foods = top_famous_foods.merge(popular_food_cities, on='famous_food')

    # Insights: Average rating and cost for top famous foods
    avg_rating_cost = df.groupby('famous_food').agg(
        avg_rating=('rating', 'mean'),
        avg_cost_for_two=('cost_for_two', 'mean')
    ).reset_index()

    # Filter for only the top 10 famous foods
    avg_rating_cost = avg_rating_cost[avg_rating_cost['famous_food'].isin(top_famous_foods['famous_food'])]

    # Merge all information into a single DataFrame
    detailed_famous_foods = top_famous_foods.merge(avg_rating_cost, on='famous_food')

    # Convert to dictionary format for JSON response
    ff = {
        'detailed_famous_foods': detailed_famous_foods.to_dict(orient='records')
    }

    return ff
